docs/blog/posts/darren-year-in-review.md

```diff
@@ -0,0 +1,171 @@
+
+The project is available [on GitHub](https://github.com/darrenburns/kupo).
+
+## Better diffs in the terminal
+
+Diffs in the terminal are often difficult to read at a glance. I wanted to see how close I could get to achieving a diff display of a quality similar to that found in the GitHub UI.
+
+To attempt this, I built a tool called [Dunk](https://github.com/darrenburns/dunk). It’s a command line program which you can pipe your `git diff` output into, and it’ll convert it into something which I find much more readable.
+
+![Untitled](../images/darren-year-in-review/Untitled%201.png)
+
+Although I’m not particularly proud of the code - there are a lot of “hacks” going on, but I’m proud of the result. If anything, it shows what can be achieved for tools like this.
+
+For many diffs, the difference between running `git diff` and `git diff | dunk | less -R` is night and day.
+
+![Untitled](../images/darren-year-in-review/Untitled%202.png)
+
+It’d be interesting to revisit this at some point.
+It has its issues, but I’d love to see how it can be used alongside Textual to build a terminal-based diff/merge tool. Perhaps it could be combined with…
+
+## Code editor floating gutter
+
+This is a common feature in text editors and IDEs: when you scroll to the right, you should still be able to see what line you’re on. Out of interest, I tried to recreate the effect in the terminal using Textual.
+
+![floating-gutter.gif](../images/darren-year-in-review/floating-gutter.gif)
+
+Textual CSS offers a `dock` property which allows you to attach a widget to an edge of its parent.
+By creating a widget that contains a vertical list of numbers and setting the `dock` property to `left`, we can create a floating gutter effect.
+Then, we just need to keep the `scroll_y` in sync between the gutter and the content to ensure the line numbers stay aligned.
+
+## Dropdown autocompletion menu
+
+While working on [Shira](https://github.com/darrenburns/shira) (a proof-of-concept, terminal-based Python object explorer), I wrote some autocompleting dropdown functionality.
+
+![shira-demo.gif](../images/darren-year-in-review/shira-demo.gif)
+
+Textual forgoes the z-index concept from browser CSS and instead uses a “named layer” system. Using the `layers` property you can defined an ordered list of named layers, and using the `layer` property, you can assign a descendant widget to one of those layers.
+
+By creating a new layer above all others and assigning a widget to that layer, we can ensure that widget is painted above everything else.
+
+In order to determine where to place the dropdown, we can track the current value in the dropdown by `watch`ing the reactive input “value” inside the Input widget. This method will be called every time the `value` of the Input changes, and we can use this hook to amend the position of our dropdown position to accommodate for the length of the input value.
```

Comment:
I feel this part might be a bit jarring - I randomly start going into a bit too much detail. Maybe I should talk a bit more about Shira and the functionality of the autocomplete library at a high level instead.

Reply:
Think I'll be publishing on Sunday. I think this is great, but feel free to PR edits before then if you like.

Reply:
@willmcgugan Should I set `draft: false` now or will you do it?

Reply:
I'll do it.